File: 17.letter-combinations-of-a-phone-number.cs
/*
 * @lc app=leetcode id=17 lang=csharp
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution {
    private Dictionary<char, string> hashTable = new Dictionary<char, string> {
        {'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"},
        {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}
    };

    private List<string> result = new List<string>();
    private string digits;

    private void DFS(int pos, string currentString) {
        if (pos == digits.Length) {
            if (!string.IsNullOrEmpty(currentString)) {
                result.Add(currentString);
            }
        }
        else {
            string s = hashTable[digits[pos]];
            foreach (char c in s) {
                DFS(pos + 1, currentString + c);
            }
        }
    }

    public IList<string> LetterCombinations(string digits) {
        result.Clear();
        this.digits = digits;
        DFS(0, "");
        return result;
    }
}

// @lc code=end

File: 17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python3
#
# [17] Letter Combinations of a Phone Number
#

# @lc code=start
from typing import List


class Solution:
    def dfs(self, pos, string):
        if pos == len(self.digits):
            if string: 
                self.result.append(string)
        else:
            s = self.hash_table[self.digits[pos]]
            for c in s:
                self.dfs(pos + 1, string + c)
        
    def letterCombinations(self, digits: str) -> List[str]:
        self.hash_table ={"2": "abc","3":"def", "4":"ghi", "5":"jkl" , "6": "mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}
        
        self.result = []
        self.digits = digits
        
        self.dfs(0,'')
        return self.result
# @lc code=end

